### PR TITLE
Refactor do post

### PR DESCRIPTION
**Pull request by annyhe**
- please comment on the refactor of doPost: the intent is to extract functions from doPost, for  testability
- will add tests

### CODE REVIEW

**Comment by annyhe** [comment]
contains the changes from #534

File: api/v1/helpers/verbs/postUtils.js [hunk 1]
```diff
@@ -0,0 +1,138 @@
+/**
+ * Copyright (c) 2016, salesforce.com, inc.
```
Comment (iamigo) [inline api/v1/helpers/verbs/postUtils.js]: 2017

File: api/v1/helpers/verbs/postUtils.js [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or
+ * https://opensource.org/licenses/BSD-3-Clause
+ */
+
+/**
+ * api/v1/helpers/verbs/postUtils.js
+ */
+'use strict';
+const NOT_FOUND = -1;
```
Comment (shriramshankar) [inline api/v1/helpers/verbs/postUtils.js]: unused variable

File: api/v1/helpers/verbs/postUtils.js [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+const logAPI = require('../../../../utils/apiLog').logAPI;
+const publisher = require('../../../../realtime/redisPublisher');
+const realtimeEvents = require('../../../../realtime/constants').events;
+const authUtils = require('../authUtils');
+const u = require('./utils');
+const featureToggles = require('feature-toggles');
+
+/**
+ * @param {Object} params From swagger
+ * @param {Object} props From express
```
Comment (shriramshankar) [inline api/v1/helpers/verbs/postUtils.js]: "props"  is not from express. I think it is the model on which the post action is happening.

File: api/v1/helpers/verbs/postUtils.js [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    return (props.modelName === 'Sample') ?
+      u.createSample(req, props) : props.model.create(toPost);
+  }
+}
+
+/**
+ *
+ * @param {Object} o Sequelize object
+ * @param {Object} resultObj For logging
+ * @param {Object} props From express
```
Comment (shriramshankar) [inline api/v1/helpers/verbs/postUtils.js]: same comment for props as above.

File: api/v1/helpers/verbs/postUtils.js [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+      // if no user found, proceed with post sample
+      if (err.status === constants.httpStatus.FORBIDDEN) {
+        return isCacheOnAndIsSample ?
+          redisModelSample.postSample(params, false) :
+          props.model.create(toPost);
+      }
+
+      // non FORBIDDEN error. Throw it to be caught by the latter .catch.
+      // this bypasses the postPromise.then function
```
Comment (shriramshankar) [inline api/v1/helpers/verbs/postUtils.js]: /*
*
*/
for multi line comments

File: api/v1/helpers/verbs/postUtils.js [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+}
+
+/**
+ *
+ * @param {Object} o Sequelize object
+ * @param {Object} resultObj For logging
+ * @param {Object} props From express
+ * @param {Object} res From express
+ */
+function handlePostResult(o, resultObj, props, res, req) {
```
Comment (shriramshankar) [inline api/v1/helpers/verbs/postUtils.js]: req param definition is missing.

File: api/v1/helpers/verbs/doPost.js [hunk 7]
```diff
@@ -32,94 +26,10 @@ const featureToggles = require('feature-toggles');
...
-    // if response directly from sequelize, call reload to attach
-    // the associations
-    if (featureToggles.isFeatureEnabled('returnUser') && o.get) {
-      o.reload()
-      .then(() => res.status(httpStatus.CREATED).json(
-          u.responsify(o, props, req.method)));
-    } else {
-      return res.status(httpStatus.CREATED).json(u.responsify(o, props, req.method));
-    }
-  })
```
Comment (shriramshankar) [inline api/v1/helpers/verbs/doPost.js]: Good to see these go away.

**Comment by annyhe** [comment]
@iamigo @shriramshankar Updated as per comments. Please review

File: tests/api/v1/verbs/doPost.js [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+ * Copyright (c) 2017, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or
+ * https://opensource.org/licenses/BSD-3-Clause
+ */
+
+/**
+ * tests/api/v1/verbs/doPost.js
+ */
```
Comment (shriramshankar) [inline tests/api/v1/verbs/doPost.js]: Looks like the tests are for the functions in "api/v1/helpers/verbs/postUtils.js". Shouldn't you name this file to match that?

Comment (annyhe) [reply tests/api/v1/verbs/doPost.js]: good point. will change.

File: api/v1/helpers/verbs/postUtils.js [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+  // if cache is on, check relatedLinks
+  if (isCacheOnAndIsSample) {
+    const rLinks = toPost.relatedLinks;
+    if (rLinks) {
+      u.checkDuplicateRLinks(rLinks);
+    }
+
+    // since cache is on AND get user.
+    // populate the user object.
+    // need to pass down the user id to populate provider field
```
Comment (shriramshankar) [inline api/v1/helpers/verbs/postUtils.js]: @annyhe
 /*
 *
*/
for multiline comments please. We can fix it later.

**Pull request merged**